### DISCUSSION TITLE
Is cyclelayout keybind with no arguments not supposed to work?

### DISCUSSION THREAD

**Discussion by 0xGodspeed** [bakkeby/dwm-flexipatch#358] [Q&A] [answered]
`{ MODKEY,                       XK_grave,      cyclelayout, 	       { } },`

Earlier it used to cycle through the layouts but after updating to the latest version the bind doesn't work.

**Comment by bakkeby** [comment, marked as answer]
Looking at the original [cyclelayouts](https://dwm.suckless.org/patches/cyclelayouts/) patch we have that the argument passed to the `cyclelayout` function is supposed to indicate direction.

```c
   { MODKEY|ControlMask,           XK_comma,  cyclelayout,    {.i = -1 } },
   { MODKEY|ControlMask,           XK_period, cyclelayout,    {.i = +1 } },
```

but the function itself just checked for i > 0:

```c
cyclelayout(const Arg *arg) {
   Layout *l;
   for(l = (Layout *)layouts; l != selmon->lt[selmon->sellt]; l++);
   if(arg->i > 0) {
       if(l->symbol && (l + 1)->symbol)
           setlayout(&((Arg) { .v = (l + 1) }));
       else
           setlayout(&((Arg) { .v = layouts }));
   } else {
       if(l != layouts && (l - 1)->symbol)
           setlayout(&((Arg) { .v = (l - 1) }));
       else
           setlayout(&((Arg) { .v = &layouts[LENGTH(layouts) - 2] }));
   }
}
```

So passing no argument (defaults to 0) does not make much sense in this context, but has the effect that it would do a reverse cycle of layouts.

The patch also added a `NULL` layout in order for the `cyclelayout` function to know when it has reached the end of the layouts array.

```diff
@@ -41,6 +41,7 @@ static const Layout layouts[] = {
    { "[]=",      tile },    /* first entry is default */
    { "><>",      NULL },    /* no layout function means floating behavior */
    { "[M]",      monocle },
+   { NULL,       NULL },
 };
```

This `NULL` layout is unnecessary and caused some issues like #331. To address that the patch was refactored and simplified with commit [f713dde](https://github.com/bakkeby/dwm-flexipatch/commit/f713ddee39070ce85b5bad1ac20af497f8fae549). Passing an empty argument (0) will now just change back to the current layout. This won't have any effect unless you are using flextile deluxe, in which case some layout settings may be reset.

Does that answer your question?

**Reply by 0xGodspeed** [reply]
It does. Thanks.
